temporalio/worker.py

```diff
@@ -1098,17 +1092,19 @@ def register_heartbeater(
             self._queue_poller_executor.submit(self._heartbeat_processor)
         return _MultiprocessingSharedHeartbeatSender(self._heartbeat_queue)
 
-    def unregister_heartbeater(self, task_token: bytes) -> None:
-        # Put a completion on the queue and wait for it to happen
-        flush_complete = threading.Event()
-        self._heartbeat_completions[task_token] = flush_complete.set
+    async def unregister_heartbeater(self, task_token: bytes) -> None:
+        # Put a callback on the queue and wait for it to happen
+        loop = asyncio.get_running_loop()
+        finish_event = asyncio.Event()
+        self._heartbeat_completions[task_token] = lambda: loop.call_soon_threadsafe(
+            finish_event.set
+        )
         try:
-            # 30 seconds to put complete, 30 to get notified should be plenty
+            # We only give the queue a few seconds to have enough room
             self._heartbeat_queue.put(
-                (task_token, _multiprocess_heartbeat_complete), True, 30
+                (task_token, _multiprocess_heartbeat_complete), True, 5
```

Comment:
nit: instead of this string you can use `object()` to create a unique empty object.